packages/dev/s2-docs/src/client.tsx

```diff
@@ -30,14 +30,52 @@ let updateRoot = hydrate({
 let currentNavigationId = 0;
 let currentAbortController: AbortController | null = null;
 
+// Store scroll positions by pathname
+const scrollPositions = new Map<string, {scrollTop: number, windowScrollTop: number}>();
+
+// Track the current pathname so we can save scroll position on popstate
+// (when popstate fires, location.pathname has already changed to the destination)
+let currentPathname = location.pathname;
+
+function getScrollContainer(): HTMLElement | null {
+  return document.querySelector('main');
+}
+
+// Save scroll position for a given pathname
+function saveScrollPosition(pathname: string) {
+  let scrollContainer = getScrollContainer();
+  let scrollTop = scrollContainer?.scrollTop ?? 0;
+  let windowScrollTop = window.scrollY;
+  scrollPositions.set(pathname, {scrollTop, windowScrollTop});
+}
+
+function restoreScrollPosition(pathname: string) {
+  let position = scrollPositions.get(pathname);
+  if (position) {
+    requestAnimationFrame(() => {
+      let scrollContainer = getScrollContainer();
+      if (scrollContainer) {
+        scrollContainer.scrollTop = position.scrollTop;
+      }
+      window.scrollTo(0, position.windowScrollTop);
```

Comment:
Wouldn't it need to be the inner div that we scroll not the window?

Reply:
That one is for the mobile layout.